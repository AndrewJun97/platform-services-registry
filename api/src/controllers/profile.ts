//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import { Contact } from '../db/model/contact';
import { ProjectProfile } from '../db/model/profile';
import { QuotaSize } from '../db/model/quota';
<<<<<<< HEAD
import { Request } from '../db/model/request';
import { getQuotaSize } from '../libs/profile';
import { getAllowedQuotaSizes } from '../libs/quota';
import { requestProfileContactsEdit, requestProfileQuotaSizeEdit } from '../libs/request';
import shared from '../libs/shared';
import { validateRequiredFields } from '../libs/utils';
=======
import { getProfileCurrentQuotaSize } from '../libs/profile';
import { getAllowedQuotaSizes } from '../libs/quota';
import { fetchEditRequests, requestContactsEdit, requestQuotaSizeEdit } from '../libs/request';
import shared from '../libs/shared';
>>>>>>> 5fc9710 (refactor and modify unit tests)

const dm = new DataManager(shared.pgPool);

export const addContactToProfile = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId, contactId } = params;

  try {
    await ProfileModel.addContactToProfile(Number(profileId), Number(contactId));

<<<<<<< HEAD
    res.status(201).end();
=======
    res.status(204).end();
>>>>>>> 5fc9710 (refactor and modify unit tests)
  } catch (err) {
    const message = `Unable to add contact to profile`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileContacts = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;
  const { profileId } = params;

  try {
    const projectContacts = await ContactModel.findForProject(Number(profileId));

    res.status(200).json(projectContacts);
  } catch (err) {
    if (err.code) {
      throw err;
    }

    const message = `Unable fetch profile contacts with profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProfileContacts = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;
  const { profileId } = params;
  const { productOwner, technicalContact } = body;
<<<<<<< HEAD
  const contacts = [productOwner, technicalContact];

  // TODO:(yh) add more data sanity check
  // check the passed contacts have no dupliates
  // check contact_id is associated with the queried profile_id
  // check role_id points to the legit role TC / PO
  contacts.forEach((contact: Contact): void => {
    const rv = validateRequiredFields(ContactModel.requiredFields.concat(['id']), contact);
    if (rv) {
      throw rv;
    }
  });

  try {
    const currentContacts: Contact[] = await ContactModel.findForProject(Number(profileId));

    const editCompares: boolean[] = [];
    contacts.forEach((contact: Contact): void => {
      const currentContact = currentContacts.filter(cc => cc.id === contact.id).pop();
      if (!currentContact) {
        throw new Error('Cant get current contact');
      }
      editCompares.push(currentContact.githubId !== contact.githubId);
      editCompares.push(currentContact.email !== contact.email);
    });

    const provisionerRelatedChanges = editCompares.some(editCompare => editCompare);
    if (provisionerRelatedChanges) {
      await requestProfileContactsEdit(Number(profileId), contacts);
      res.status(202).end();
    } else {
      const contactPromises = contacts.map((contact: Contact) => {
        if (!contact.id) {
          throw new Error('Cant get contact id');
        }
        return ContactModel.update(contact.id, contact);
      });
      await Promise.all(contactPromises);
      res.status(204).end();
    }
=======

  try {
    const contacts = [productOwner, technicalContact];
    const currentPOvalues = await ContactModel.findById(productOwner.id);
    const currentTCvalues = await ContactModel.findById(technicalContact.id);

    const editCompares = [
      currentPOvalues.githubId !== productOwner.githubId,
      currentPOvalues.email !== productOwner.email,
      currentTCvalues.githubId !== technicalContact.githubId,
      currentTCvalues.email !== technicalContact.email,
    ];
    const provisionerRelatedChanges = editCompares.some(editCompare => editCompare);

    if (provisionerRelatedChanges) {
      await requestContactsEdit(Number(profileId), body);
    } else {
      const updatePromises: any = [];
      contacts.forEach((contact: Contact) => {
        updatePromises.push(ContactModel.update(Number(contact.id), contact));
      });

      await Promise.all(updatePromises);
    }

    res.status(204).end();
>>>>>>> 5fc9710 (refactor and modify unit tests)
  } catch (err) {
    const message = `Unable to update contacts with profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileQuotaSize = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const profile: ProjectProfile = await ProfileModel.findById(Number(profileId));
<<<<<<< HEAD
    const quotaSize: QuotaSize = await getQuotaSize(profile);
=======
    const quotaSize: QuotaSize = await getProfileCurrentQuotaSize(profile);
>>>>>>> 5fc9710 (refactor and modify unit tests)

    res.status(200).json(quotaSize);
  } catch (err) {
    if (err.code) {
      throw err;
    }

    const message = `Unable fetch profile quota size with profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileAllowedQuotaSizes = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const profile: ProjectProfile = await ProfileModel.findById(profileId);
<<<<<<< HEAD
    const quotaSize: QuotaSize = await getQuotaSize(profile);
    const allowedQuotaSizes: QuotaSize[] = getAllowedQuotaSizes(quotaSize);

    res.status(200).json(allowedQuotaSizes);
  } catch (err) {
    const message = `Unable to fetch allowed quota-sizes for profile ${profileId}`;
=======
    const quotaSize: QuotaSize = await getProfileCurrentQuotaSize(profile);
    const results = getAllowedQuotaSizes(quotaSize);

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable to fetch quota options for profile ${profileId}`;
>>>>>>> 5fc9710 (refactor and modify unit tests)
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProfileQuotaSize = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;
  const { requestedQuotaSize } = body;

  try {
    const profile: ProjectProfile = await ProfileModel.findById(profileId);
<<<<<<< HEAD
    const quotaSize: QuotaSize = await getQuotaSize(profile);
    const allowedQuotaSizes: QuotaSize[] = getAllowedQuotaSizes(quotaSize);

    // verify if requested quota size is valid
    if (!(requestedQuotaSize && allowedQuotaSizes.includes(requestedQuotaSize))) {
      throw new Error('Please provide correct requested quota size in body');
    }

    await requestProfileQuotaSizeEdit(Number(profileId), requestedQuotaSize);

    res.status(204).end();
  } catch (err) {
    const message = `Unable to update quota-size for profile ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileEditRequests = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { RequestModel } = dm;
  const { profileId } = params;

  try {
    const editRequests: Request[] = await RequestModel.findForProfile(Number(profileId));

    res.status(200).json(editRequests);
  } catch (err) {
=======
    const quotaSize: QuotaSize = await getProfileCurrentQuotaSize(profile);
    const allowedQuotaSizes = getAllowedQuotaSizes(quotaSize);

    // verify if requested quota size is valid
    if (!(requestedQuotaSize && allowedQuotaSizes.includes(requestedQuotaSize))) {
      const errmsg = 'Please provide correct requested quota size in body';
      throw new Error(errmsg);
    }

    await requestQuotaSizeEdit(Number(profileId), body);

    res.status(204).end();
  } catch (err) {
    const message = `Unable to request quota size for profile ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileEditRequests = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { profileId } = params;

  try {
    const results = await fetchEditRequests(profileId);

    res.status(200).json(results);
  } catch (err) {
>>>>>>> 5fc9710 (refactor and modify unit tests)
    if (err.code) {
      throw err;
    }

<<<<<<< HEAD
    const message = `Unable to fetch profile edit requests with profile ID ${profileId}`;
=======
    const message = `Unable fetch requests with profile ID ${profileId}`;
>>>>>>> 5fc9710 (refactor and modify unit tests)
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
