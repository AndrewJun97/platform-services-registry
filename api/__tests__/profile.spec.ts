//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import fs from 'fs';
import path from 'path';
import { Pool } from 'pg';
import {
  addContactToProfile,
  fetchProfileAllowedQuotaSizes,
  fetchProfileContacts,
  fetchProfileEditRequests,
  fetchProfileQuotaSize,
  updateProfileContacts,
  updateProfileQuotaSize
} from '../src/controllers/profile';
import ContactModel from '../src/db/model/contact';
import { getProfileCurrentQuotaSize } from '../src/libs/profile';
import { getAllowedQuotaSizes } from '../src/libs/quota';
import { fetchEditRequests, requestContactsEdit, requestQuotaSizeEdit } from '../src/libs/request';
import FauxExpress from './src/fauxexpress';

const p0 = path.join(__dirname, 'fixtures/select-profile.json');
const selectProfile = JSON.parse(fs.readFileSync(p0, 'utf8'));

const p1 = path.join(__dirname, 'fixtures/select-profile-contacts.json');
const selectProfilesContacts = JSON.parse(fs.readFileSync(p1, 'utf8'));
const productOwner = selectProfilesContacts[0];
const technicalContact = selectProfilesContacts[1];

const client = new Pool().connect();

jest.mock('../src/libs/profile', () => {
  return {
    getProfileCurrentQuotaSize: jest.fn().mockResolvedValue('small'),
  };
});

jest.mock('../src/libs/quota', () => {
  return {
    getAllowedQuotaSizes: jest.fn().mockReturnValue(['medium']),
  };
});

jest.mock('../src/libs/request', () => ({
  requestContactsEdit: jest.fn(),
  requestQuotaSizeEdit: jest.fn(),
  fetchEditRequests: jest.fn(),
}));

describe('Profile event handlers', () => {
  let ex;

  beforeEach(() => {
    jest.clearAllMocks();

    ex = new FauxExpress();
  });

  it('Link a contact to profile successfully', async () => {
    const req = {
      params: { profileId: 4, contactId: 1 },
    };

    client.query.mockReturnValueOnce({ rows: [] });

    await addContactToProfile(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
  });

  it('Link a contact to profile should throw', async () => {
    const req = {
      params: { profileId: 4, contactId: 1 },
    };

    client.query.mockImplementation(() => { throw new Error() });

    await expect(addContactToProfile(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.responseData).toBeUndefined();
  });

  it('A profiles contacts are returned', async () => {
    const req = {
      params: { profileId: 1 },
    };

    client.query.mockReturnValueOnce({ rows: selectProfilesContacts });

    await fetchProfileContacts(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('Fetch single profile contacts should throw', async () => {
    const req = {
      params: { profileId: 1 },
    };
    client.query.mockImplementation(() => { throw new Error() });

    await expect(fetchProfileContacts(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.responseData).toBeUndefined();
  });

  it('Update profile contacts with non provisioner-related changes', async () => {
    const body = {
      productOwner: {
        id: 1,
        firstName: 'JaneTEST',
        lastName: 'DoeTEST',
        email: 'jane@example.com',
        githubId: 'jane1100',
        roleId: 1,
      },
      technicalContact: {
        id: 2,
        firstName: 'JimTEST',
        lastName: 'DoeTEST',
        email: 'jim@example.com',
        githubId: 'jim1100',
        roleId: 2,
      }
    };
    const req = {
      params: { profileId: 4 },
      body,
    };

    client.query.mockReturnValueOnce({ rows: [productOwner] });
    client.query.mockReturnValueOnce({ rows: [technicalContact] });
    const update = ContactModel.prototype.update = jest.fn();

    await updateProfileContacts(req, ex.res);
    expect(update).toHaveBeenCalledTimes(2);
    expect(requestContactsEdit).toHaveBeenCalledTimes(0);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
  });

  it('Request profile contacts edit with provisioner-related changes', async () => {
    const body = {
      productOwner: {
        id: 1,
        firstName: 'Jane',
        lastName: 'Doe',
        email: 'janeTEST@example.com',
        githubId: 'jane1100TEST',
        roleId: 1,
      },
      technicalContact: {
        id: 2,
        firstName: 'Jim',
        lastName: 'Doe',
        email: 'jimTEST@example.com',
        githubId: 'jim1100TEST',
        roleId: 2,
      }
    };
    const req = {
      params: { profileId: 4 },
      body,
    };

    client.query.mockReturnValueOnce({ rows: [productOwner] });
    client.query.mockReturnValueOnce({ rows: [technicalContact] });
    const update = ContactModel.prototype.update = jest.fn();

    await updateProfileContacts(req, ex.res);
    expect(update).toHaveBeenCalledTimes(0);
    expect(requestContactsEdit).toHaveBeenCalledTimes(1);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
  });

  it('Update profile contacts should throw', async () => {
    const req = {
      params: { profileId: 4 },
      body: {},
    };

    await expect(updateProfileContacts(req, ex.res)).rejects.toThrow();

    expect(ex.res.status).not.toBeCalled();
    expect(ex.res.end).not.toBeCalled();
  });

  it('A profiles quota size is returned', async () => {
    const req = {
      params: { profileId: 4 },
    };

    client.query.mockReturnValueOnce({ rows: selectProfile });

    await fetchProfileQuotaSize(req, ex.res);
    expect(getProfileCurrentQuotaSize).toHaveBeenCalledTimes(1);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('Fetch profiles quota size should throw', async () => {
    const req = {
      params: { profileId: 4 },
    };

    client.query.mockImplementation(() => { throw new Error() });

    await expect(fetchProfileQuotaSize(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.responseData).toBeUndefined();
  });

  it('A list of quota options for a given profile is returned', async () => {
    const req = {
      params: { profileId: 4 },
    };

    client.query.mockReturnValueOnce({ rows: selectProfile });

    await fetchProfileAllowedQuotaSizes(req, ex.res);
    expect(getAllowedQuotaSizes).toHaveBeenCalledTimes(1);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('Fetch a list of quota options for a given profile should throw', async () => {
    const req = {
      params: { profileId: 4 },
    };

    client.query.mockImplementation(() => { throw new Error() });

    await expect(fetchProfileAllowedQuotaSizes(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.responseData).toBeUndefined();
  });

  it('Request profile quota size edit successfully', async () => {
    const body = {
      requestedQuotaSize: 'medium',
    };
    const req = {
      params: { profileId: 4 },
      body,
    };

    client.query.mockReturnValueOnce({ rows: selectProfile });

    await updateProfileQuotaSize(req, ex.res);
    expect(requestQuotaSizeEdit).toHaveBeenCalledTimes(1);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
  });

  it('Request profile quota size edit should throw due to invalid quota upgrade', async () => {
    const body = {
      requestedQuotaSize: 'large',
    };
    const req = {
      params: { profileId: 4 },
      body,
    };

    client.query.mockReturnValueOnce({ rows: selectProfile });

    await expect(updateProfileQuotaSize(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();
    expect(requestQuotaSizeEdit).toHaveBeenCalledTimes(0);
    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.responseData).toBeUndefined();
  });

  it('Request profile quota size edit should throw due to db transaction issue', async () => {
    const body = {
      requestedQuotaSize: 'medium',
    };
    const req = {
      params: { profileId: 4 },
      body,
    };

    client.query.mockImplementation(() => { throw new Error() });

    await expect(updateProfileQuotaSize(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.responseData).toBeUndefined();
  });

  it('Profile edit requests are returned', async () => {
    const req = {
      params: { profileId: 4 },
    };

    await fetchProfileEditRequests(req, ex.res);
    expect(fetchEditRequests).toHaveBeenCalledTimes(1);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });
});
