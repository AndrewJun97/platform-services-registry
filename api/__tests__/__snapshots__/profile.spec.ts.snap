// Jest Snapshot v1, https://goo.gl/fbAQLP

<<<<<<< HEAD
exports[`Profile event handlers A list of quota options for a given profile is returned 1`] = `200`;

exports[`Profile event handlers A list of quota options for a given profile is returned 2`] = `
Array [
  "medium",
]
`;

=======
exports[`Profile event handlers A list of quota options for a given profile is returned 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM profile
          WHERE id = $1 AND archived = false;",
      "values": Array [
        4,
      ],
    },
  ],
]
`;

exports[`Profile event handlers A list of quota options for a given profile is returned 2`] = `200`;

exports[`Profile event handlers A list of quota options for a given profile is returned 3`] = `
Array [
  "medium",
]
`;

>>>>>>> 5fc9710 (refactor and modify unit tests)
exports[`Profile event handlers A profiles contacts are returned 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT contact.*
          FROM contact
          JOIN profile_contact ON contact.id = profile_contact.contact_id
            WHERE profile_contact.profile_id = $1;
      ",
      "values": Array [
        1,
      ],
    },
  ],
]
`;

exports[`Profile event handlers A profiles contacts are returned 2`] = `200`;

exports[`Profile event handlers A profiles contacts are returned 3`] = `
Array [
  Object {
    "archived": false,
    "createdAt": "2020-09-10T18:14:13.436Z",
    "email": "jane@example.com",
    "firstName": "Jane",
    "githubId": "jane1100",
    "id": 1,
    "lastName": "Doe",
    "roleId": 1,
    "updatedAt": "2020-09-10T18:14:13.436Z",
  },
  Object {
    "archived": false,
    "createdAt": "2020-09-10T18:14:13.436Z",
    "email": "john.doe@example.com",
    "firstName": "John",
    "githubId": "john1100",
    "id": 2,
    "lastName": "Doe",
    "roleId": 2,
    "updatedAt": "2020-09-10T18:14:13.436Z",
  },
]
`;

<<<<<<< HEAD
exports[`Profile event handlers A profiles quota size is returned 1`] = `200`;

exports[`Profile event handlers A profiles quota size is returned 2`] = `"small"`;

exports[`Profile event handlers Fetch a list of quota options for a given profile should throw 1`] = `"Unable to fetch allowed quota-sizes for profile 4"`;

exports[`Profile event handlers Fetch profiles quota size should throw 1`] = `"Unable fetch profile quota size with profile ID 4"`;

exports[`Profile event handlers Fetch single profile contacts should throw 1`] = `"Unable fetch profile contacts with profile ID 1"`;

exports[`Profile event handlers Fetch single profile contacts should throw 2`] = `
=======
exports[`Profile event handlers A profiles quota size is returned 1`] = `
>>>>>>> 5fc9710 (refactor and modify unit tests)
Array [
  Array [
    Object {
      "text": "
<<<<<<< HEAD
        SELECT contact.*
          FROM contact
          JOIN profile_contact ON contact.id = profile_contact.contact_id
            WHERE profile_contact.profile_id = $1;
      ",
      "values": Array [
        1,
=======
        SELECT * FROM profile
          WHERE id = $1 AND archived = false;",
      "values": Array [
        4,
      ],
    },
  ],
]
`;

exports[`Profile event handlers A profiles quota size is returned 2`] = `200`;

exports[`Profile event handlers A profiles quota size is returned 3`] = `"small"`;

exports[`Profile event handlers Fetch a list of quota options for a given profile should throw 1`] = `"Unable to fetch quota options for profile 4"`;

exports[`Profile event handlers Fetch a list of quota options for a given profile should throw 2`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM profile
          WHERE id = $1 AND archived = false;",
      "values": Array [
        4,
>>>>>>> 5fc9710 (refactor and modify unit tests)
      ],
    },
  ],
]
`;

<<<<<<< HEAD
exports[`Profile event handlers Link a contact to profile should throw 1`] = `"Unable to add contact to profile"`;
=======
exports[`Profile event handlers Fetch profiles quota size should throw 1`] = `"Unable fetch profile quota size with profile ID 4"`;

exports[`Profile event handlers Fetch profiles quota size should throw 2`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM profile
          WHERE id = $1 AND archived = false;",
      "values": Array [
        4,
      ],
    },
  ],
]
`;
>>>>>>> 5fc9710 (refactor and modify unit tests)

exports[`Profile event handlers Link a contact to profile should throw 2`] = `
Array [
  Array [
    Object {
      "text": "
        INSERT INTO profile_contact
          (profile_id, contact_id)
          VALUES ($1, $2) RETURNING *;",
      "values": Array [
        4,
        1,
      ],
    },
  ],
]
`;

exports[`Profile event handlers Link a contact to profile successfully 1`] = `
Array [
  Array [
    Object {
      "text": "
        INSERT INTO profile_contact
          (profile_id, contact_id)
          VALUES ($1, $2) RETURNING *;",
      "values": Array [
        4,
        1,
      ],
    },
  ],
]
`;

<<<<<<< HEAD
exports[`Profile event handlers Link a contact to profile successfully 2`] = `201`;

exports[`Profile event handlers Link a contact to profile successfully 3`] = `undefined`;

exports[`Profile event handlers Profile edit requests are returned 1`] = `200`;

exports[`Profile event handlers Profile edit requests are returned 2`] = `
Array [
  Object {
    "archived": false,
    "createdAt": "2020-09-10T18:14:13.304Z",
    "editObject": Object {
      "quota": "medium",
      "quotas": Object {
        "cpu": Object {
          "limits": "16",
          "requests": "8",
        },
        "memory": Object {
          "limits": "64Gi",
          "requests": "32Gi",
        },
        "storage": Object {
          "backup": "50Gi",
          "block": "100Gi",
          "capacity": "100Gi",
          "file": "100Gi",
        },
      },
    },
    "editType": "quotaSize",
    "id": 1,
    "natsContext": "nats_context_stub",
    "natsSubject": "registry_project_provisioning",
    "profileId": 1,
    "updatedAt": "2020-09-10T18:14:13.304Z",
  },
  Object {
    "archived": false,
    "createdAt": "2020-09-10T18:14:13.304Z",
    "editObject": Array [
      Object {
        "email": "jane@example.com",
        "firstName": "Jane",
        "githubId": "janedoe",
        "id": 1,
        "lastName": "Doe",
        "roleId": 1,
      },
      Object {
        "email": "jim@example.com",
        "firstName": "Jim",
        "githubId": "jim1100",
        "id": 2,
        "lastName": "Doe",
        "roleId": 2,
      },
    ],
    "editType": "contacts",
    "id": 2,
    "natsContext": "nats_context_stub",
    "natsSubject": "registry_project_provisioning",
    "profileId": 4,
    "updatedAt": "2020-09-10T18:14:13.304Z",
  },
]
`;

exports[`Profile event handlers Request profile contacts edit with provisioner-related changes 1`] = `202`;

exports[`Profile event handlers Request profile contacts edit with provisioner-related changes 2`] = `undefined`;

exports[`Profile event handlers Request profile quota size edit should throw due to db transaction issue 1`] = `"Unable to update quota-size for profile 4"`;

exports[`Profile event handlers Request profile quota size edit should throw due to invalid quota upgrade 1`] = `"Unable to update quota-size for profile 4"`;

exports[`Profile event handlers Request profile quota size edit successfully 1`] = `204`;

exports[`Profile event handlers Request profile quota size edit successfully 2`] = `undefined`;

exports[`Profile event handlers Update profile contacts with non provisioner-related changes 1`] = `204`;

exports[`Profile event handlers Update profile contacts with non provisioner-related changes 2`] = `undefined`;
=======
exports[`Profile event handlers Link a contact to profile should throw 1`] = `"Unable to create contact"`;

exports[`Profile event handlers Link a contact to profile should throw 2`] = `
Array [
  Array [
    Object {
      "text": "
        INSERT INTO profile_contact
          (profile_id, contact_id)
          VALUES ($1, $2) RETURNING *;",
      "values": Array [
        4,
        1,
      ],
    },
  ],
]
`;

exports[`Profile event handlers Link a contact to profile successfully 1`] = `
Array [
  Array [
    Object {
      "text": "
        INSERT INTO profile_contact
          (profile_id, contact_id)
          VALUES ($1, $2) RETURNING *;",
      "values": Array [
        4,
        1,
      ],
    },
  ],
]
`;

exports[`Profile event handlers Link a contact to profile successfully 2`] = `204`;

exports[`Profile event handlers Link a contact to profile successfully 3`] = `undefined`;

exports[`Profile event handlers Profile edit requests are returned 1`] = `Array []`;

exports[`Profile event handlers Profile edit requests are returned 2`] = `200`;

exports[`Profile event handlers Profile edit requests are returned 3`] = `undefined`;

exports[`Profile event handlers Request profile contacts edit with provisioner-related changes 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM contact
          WHERE id = $1 AND archived = false;",
      "values": Array [
        1,
      ],
    },
  ],
  Array [
    Object {
      "text": "
        SELECT * FROM contact
          WHERE id = $1 AND archived = false;",
      "values": Array [
        2,
      ],
    },
  ],
]
`;

exports[`Profile event handlers Request profile contacts edit with provisioner-related changes 2`] = `204`;

exports[`Profile event handlers Request profile contacts edit with provisioner-related changes 3`] = `undefined`;

exports[`Profile event handlers Request profile quota size edit should throw due to db transaction issue 1`] = `"Unable to request quota size for profile 4"`;

exports[`Profile event handlers Request profile quota size edit should throw due to db transaction issue 2`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM profile
          WHERE id = $1 AND archived = false;",
      "values": Array [
        4,
      ],
    },
  ],
]
`;

exports[`Profile event handlers Request profile quota size edit should throw due to invalid quota upgrade 1`] = `"Unable to request quota size for profile 4"`;

exports[`Profile event handlers Request profile quota size edit should throw due to invalid quota upgrade 2`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM profile
          WHERE id = $1 AND archived = false;",
      "values": Array [
        4,
      ],
    },
  ],
]
`;

exports[`Profile event handlers Request profile quota size edit successfully 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM profile
          WHERE id = $1 AND archived = false;",
      "values": Array [
        4,
      ],
    },
  ],
]
`;

exports[`Profile event handlers Request profile quota size edit successfully 2`] = `204`;

exports[`Profile event handlers Request profile quota size edit successfully 3`] = `undefined`;

exports[`Profile event handlers Update profile contacts with non provisioner-related changes 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM contact
          WHERE id = $1 AND archived = false;",
      "values": Array [
        1,
      ],
    },
  ],
  Array [
    Object {
      "text": "
        SELECT * FROM contact
          WHERE id = $1 AND archived = false;",
      "values": Array [
        2,
      ],
    },
  ],
]
`;

exports[`Profile event handlers Update profile contacts with non provisioner-related changes 2`] = `204`;

exports[`Profile event handlers Update profile contacts with non provisioner-related changes 3`] = `undefined`;
>>>>>>> 5fc9710 (refactor and modify unit tests)
